includes/discount-functions.php

```diff
@@ -373,31 +373,50 @@ function edd_get_discount_notes( $discount_id = 0 ) {
  *
  * @since 1.0
  * @since 3.0 Updated to be more efficient and make direct calls to the EDD_Discount object.
+ * @since 3.1.0.3 Added Sample size and Refresh parameters to be able to cache the result as a transient.
+ *
+ * @param int  $sample_size The number of discount codes to query. Defaults to 10 for performance.
+ * @param bool $refresh     If we should force a refresh of the transient. Defaults to false.
  *
  * @return bool
  */
-function edd_has_active_discounts() {
+function edd_has_active_discounts( $sample_size = 10, $refresh = false ) {
+
+	// If we are not forcing a refresh of the value, check the transient.
+	if ( false === $refresh ) {
+		$cached_result = get_transient( 'edd_has_active_discounts' );
+
+		if ( false !== $cached_result  ) {
+			return boolval( $cached_result );
+		}
+	}
 
 	// Query for active discounts.
 	$discounts = edd_get_discounts( array(
-		'number' => 10,
+		'number' => $sample_size,
 		'status' => 'active'
 	) );
 
 	// Bail if none.
 	if ( empty( $discounts ) ) {
-		return false;
-	}
-
-	// Check each discount for active status, applying filters, etc...
-	foreach ( $discounts as $discount ) {
-		/** @var $discount EDD_Discount */
-		if ( $discount->is_active( false, true ) ) {
-			return true;
+		$has_active_discounts = false;
+	} else {
+		// Check each discount for active status, applying filters, etc...
+		foreach ( $discounts as $discount ) {
+			/** @var $discount EDD_Discount */
+			if ( $discount->is_active( false, true ) ) {
```

Comment:
I'm wondering if the first parameter should be set to `true` if `$refresh` is true. I've manually set a lot of discounts to active even though their expiration is in the past and just loading the checkout is not getting to an active but unexpired one generally. Deactivating one discount did expire it, and the transient is set to true, but as soon as it expires, it will be false again because the other expired discounts are still active. (Granted this may be an edge case but if the transient is being forcibly refreshed with a larger sample size maybe worth doing?)